src/Components/WebAssembly/JSInterop/src/WebAssemblyJSObjectReferenceJsonConverter.cs

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.JSInterop.Implementation;
+
+namespace Microsoft.JSInterop.WebAssembly
+{
+    internal sealed class WebAssemblyJSObjectReferenceJsonConverter : JsonConverter<IJSObjectReference>
+    {
+        private readonly WebAssemblyJSRuntime _jsRuntime;
+
+        public WebAssemblyJSObjectReferenceJsonConverter(WebAssemblyJSRuntime jsRuntime)
+        {
+            _jsRuntime = jsRuntime;
+        }
+
+        public override bool CanConvert(Type typeToConvert)
+        {
+            return typeToConvert == typeof(WebAssemblyJSObjectReference) ||
+                typeToConvert == typeof(IJSObjectReference) ||
+                typeToConvert == typeof(IJSInProcessObjectReference) ||
+                typeToConvert == typeof(IJSUnmarshalledObjectReference) ||
+                typeToConvert == typeof(JSObjectReference);
```

Comment:
I don't think `JSObjectReference` should be on the list.

 * It's not needed for serialization, since the only instances that should exist on wasm are `WebAssemblyJSObjectReference`
 * It's actively harmful to allow it for deserialization, because then developers don't have to use the interface type, which means their components can no longer be tested (e.g., via bUnit).

If I remember correctly, in 5.0 we explicitly ensured we would only deserialize when the interface types were requested (and not if the concrete type was requested) for that reason.

Reply:
The current implementation specifically checks for the implementation type:

* https://github.com/dotnet/aspnetcore/blob/master/src/JSInterop/Microsoft.JSInterop/src/Infrastructure/JSObjectReferenceJsonConverter.cs#L25
* https://github.com/dotnet/aspnetcore/blob/master/src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs#L41
* `TImplementation` == `typeof(JSObjectReference)`


We could change it, but it is somewhat breaking.

Reply:
This might be an unintentional inconsistency in the implementation then. In https://github.com/dotnet/aspnetcore/blob/master/src/Shared/JSInterop/JSCallResultTypeHelper.cs, we very deliberately only included the interface types.

It's strange that you can get a concrete `JSObjectReference` as an incoming method parameter, but not as the return value  to an outgoing call. Seems clear it would be *better* if the JSON converter was tightened up to impose the same restriction, but as you've pointed out, it's somewhat breaking. 

If this was the sort of breaking change you could discover at compile-time I'd be fine with doing it in 6.0. But since instead it would show up as a hard-to-understand runtime error, I'm not really inclined to do it. I know we could theoretically implement an analyzer about this, but the cost of that seems disproportionate to the benefit.

What's your take? If we're leaving this in, it would be good at least to add a comment stating the inconsistency and that it's here for back-compat reasons even though we hope people aren't declaring `JSObjectReference` as a JSInvokable method parameter type.